### PR TITLE
Note detection open cv

### PR DESCRIPTION
**Pull request by Cryplo**

### CODE REVIEW

**Comment by Cryplo** [comment]
Looks good for me

File: autonomous.py [hunk 1]
```diff
@@ -70,6 +72,18 @@ def executeAuton(self):
                 self.taskListCounter += 1 # Move on to next task.
             return True
         
+        elif self.autonTask[0] == 'MOVE_RELATIVE':
+            x = self.autonTask[1]
```
Comment (abdelazer) [inline autonomous.py:76]: Why does MOVE call `abs()` on x but this doesn't?

File: autonomous.py [hunk 2]
```diff
@@ -70,6 +72,18 @@ def executeAuton(self):
...
+            y = self.autonTask[2]
+            bearing = self.autonTask[3]
+            if self.drivetrain.goToRelativePose(x, y, bearing):
```
Comment (abdelazer) [inline autonomous.py:79]: Similar to the above, why relative on this one and not on MOVE?

File: autonomous.py [hunk 3]
```diff
@@ -93,6 +107,35 @@ def executeAuton(self):
...
+            self.pathState = self.pathTrajectory.sample(self.autonTimer.get() - self.lastTime)
+            self.chassisSpeeds = self.holonomicController.calculateRobotRelativeSpeeds(self.swervometer.getPathPlannerPose(), self.pathState)
+            self.drivetrain.set_fwd(-self.chassisSpeeds.vy/self.maxSpeed)
+            self.drivetrain.set_strafe(self.chassisSpeeds.vx/self.maxSpeed)
+            if self.drivetrain.shouldSteerStraight():
+                if(self.team_is_red):
+                    self.drivetrain.set_rcw(self.drivetrain.steerStraight(0, 0))
+                else:
+                    self.drivetrain.set_rcw(self.drivetrain.steerStraight(0, 180))
+            if self.notedetector.hasTarget() and self.notedetector.getTargetErrorY() < 70 and self.autonTimer.get() - self.lastTime > waitTime:
```
Comment (abdelazer) [inline autonomous.py:128]: Can we pull this `70` into a variable so we know what it is and can tune it explicitly?

File: coralscript.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+from pycoral.adapters.common import input_size
+from pycoral.adapters.detect import get_objects
+from pycoral.utils.dataset import read_label_file
+from pycoral.utils.edgetpu import make_interpreter
+from pycoral.utils.edgetpu import run_inference
+from networktables import NetworkTables
+import numpy as np
+import cv2
+
+"""To Do: :)
```
Comment (abdelazer) [inline coralscript.py:10]: Is this still true?

File: autonomous.py [hunk 5]
```diff
@@ -182,6 +225,76 @@ def executeAuton(self):
                     self.drivetrain.set_rcw(0)
                     self.taskListCounter += 1
             self.drivetrain.execute('center')
+
+        # elif self.autonTask[0] == 'AIM':
```
Comment (abdelazer) [inline autonomous.py:229]: Is this code obsolete or are we close to including it?

File: autonomous.py [hunk 6]
```diff
@@ -182,6 +225,76 @@ def executeAuton(self):
...
+        #     self.mechanism.sprocketToPosition(self.mechanism.getAutoAimAngle(distance, 0))
+
+        elif self.autonTask[0] == 'POINT_TO_NOTE':
+            self.goToNote = self.autonTask[1]
+            self.backupTask = self.autonTask[2]
+            if self.notedetector.hasTarget():
+                #print(self.notedetector.getTargetErrorX())
+                if abs(self.notedetector.getTargetErrorX()) < 2.0:
+                    if self.goToNote:
+                        #self.taskList.insert(self.taskListCounter + 1, ['MOVE', self.notedetector.getTargetErrorX(), self.notedetector.getTargetErrorY() - 5, self.notedetector.getTargetErrorAngle()])
```
Comment (abdelazer) [inline autonomous.py:247]: Same: Delete comments?

File: coralscript.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        return objs[0]
+    else:
+        return False
+
+def publishBBox(obj):
+    notePub.putBoolean('isAlive', True)
+    notePub.putNumber('counter', counter)
+    print(counter)
+    if obj:
+        # notePub.putString('testKey', 'Hello world')
```
Comment (abdelazer) [inline coralscript.py:32]: Delete

**Comment by ldavidc** [comment]
Looks good to me too (we might want to keep the autonConfig in main and then add the new note detection paths manually)

File: coralscript.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+def publishVideo(image):
+    if image and notePub.getBoolean('videoRequested', False):
+        notePub.putRaw('video', image)
+    return
+
+def main():
+    print("main called")
+
+    model = "/home/mendel/notedetector29/edgetpu.tflite" # absolute path of the tflite model in the DevBoard
```
Comment (abdelazer) [inline coralscript.py:54]: Ideally these get pulled into CONSTANTS for easier adjustment

File: coralscript.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    source = 1 # default USB camera source ID is often 0, but the DevBoard likes 1
+    threshold = 0.75 # minimum confidence threshold
```
Comment (abdelazer) [inline coralscript.py:56]: Same CONSTANT suggestion

File: coralscript.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        while cap.isOpened():
+            #print("cap opened")
+            ret, frame = cap.read()
+            if not ret:
+                print('frame not received')
+                break
+
+            cv2_im = frame
+            cv2_im_rgb = cv2.cvtColor(cv2_im, cv2.COLOR_BGR2RGB)
+            # publishVideo(cv2_im_rgb.tobytes())
```
Comment (abdelazer) [inline coralscript.py:76]: Delete

File: notedetector.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
+from networktables import NetworkTables
+import math
+
+class NoteDetector:
+    def __init__ (self, config):
+        #IP of the coral
+        #192.168.101.2
+        NetworkTables.initialize(server='10.10.76.2')
```
Comment (abdelazer) [inline notedetector.py:8]: CONSTANT this both for clarity and to make it easier to adjust in the future

File: notedetector.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+from networktables import NetworkTables
+import math
+
+class NoteDetector:
+    def __init__ (self, config):
+        #IP of the coral
+        #192.168.101.2
```
Comment (abdelazer) [inline notedetector.py:7]: New IP is 10.10.76.16

File: notedetector.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        NetworkTables.initialize(server='10.10.76.2')
+        self.noteSub = NetworkTables.getTable('noteDetector')
+        self.config = config
+        self.videoRequested = False
```
Comment (abdelazer) [inline notedetector.py:11]: This stuff isn't currently working, correct? If so, I'd remove all of it

File: notedetector.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        self.lastX = -1000
+        self.lastY = -1000
+        self.lastCounter = 0
+        self.sameCounterCounter = 0
+        self.sameCounterBBox = 0
+        # self.noteSub.putString('testKey', 'stuff')
+    
+    # get the camera pixel coordinates of the bounding box
+    def getXMin(self):
+        return self.noteSub.getNumber('xmin', -1000)
```
Comment (abdelazer) [inline notedetector.py:22]: Nit: Spaces after the `return` line for formatting consistency.

File: notedetector.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+class NoteDetector:
+    def __init__ (self, config):
+        #IP of the coral
+        #192.168.101.2
+        NetworkTables.initialize(server='10.10.76.2')
+        self.noteSub = NetworkTables.getTable('noteDetector')
+        self.config = config
+        self.videoRequested = False
+        self.noteSub.putBoolean('videoRequested', self.videoRequested)
+        self.lastX = -1000
```
Comment (abdelazer) [inline notedetector.py:13]: Pull all these `-1000` into a single CONSTANT?

File: notedetector.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        # the 'counter' value on the coral should increase by 1 each cycle
+        return self.noteSub.getNumber('counter', 0)
+    def hasTarget(self):
+        if self.trustCoral():
+            return bool(self.noteSub.getBoolean('hasTarget', False))
+        else:
+            return False
+
+    def trustCoral(self):
+        #return False
```
Comment (abdelazer) [inline notedetector.py:39]: Delete

File: notedetector.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+        if self.sameCounterCounter > 3:
+            return False
+        else:
+            return True
+        
+    def testBBox(self):
+        if self.getXMax == self.lastX and self.getYMax == self.lastY and self.lastX != -1000:
+            self.sameCounterBBox += 1
+
```
Comment (abdelazer) [inline notedetector.py:59]: Nit: Delete empty line to keep if/else more readable.

File: notedetector.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+        self.lastX = self.getXMax
+        self.lastY = self.getYMax
+
+        if self.sameCounterBBox > 10:    
+            return False
+        else:
+            return True
+
+    def getVideo(self):
```
Comment (abdelazer) [inline notedetector.py:71]: As above, delete?

File: notedetector.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+        self.targetPixelX = ((self.getXMin() + self.getXMax())/2 - self.config['CAMERA_PIXELS_X']/2)
+        self.targetAngleX = (self.targetPixelX/(self.config['CAMERA_PIXELS_X']/2))*(self.config['CAMERA_FOV_X']/2)
+        self.targetErrorX = self.getTargetErrorY() * math.tan(math.radians(self.targetAngleX)) + self.config['CAMERA_OFFSET_X']
+        return self.targetErrorX
+
+    def getTargetErrorY(self):
+        # y distance to the note
+
+        #self.targetPixelZ = (self.config['CAMERA_PIXELS_Z'] - (self.getYMin() + self.getYMax())/2) - self.config['CAMERA_PIXELS_Z']/2
```
Comment (abdelazer) [inline notedetector.py:91]: Delete?

File: notedetector.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        self.targetPixelZ = (self.config['CAMERA_PIXELS_Z'] - (self.getYMax())) - self.config['CAMERA_PIXELS_Z']/2
+        self.targetAngleZ = (self.targetPixelZ/(self.config['CAMERA_PIXELS_Z']/2))*(self.config['CAMERA_FOV_Z']/2) + self.config['CAMERA_ANGLE_ABOVE_HORIZONTAL']
+        self.targetErrorY = ((self.config['CAMERA_HEIGHT'] - self.config['NOTE_HEIGHT'])/math.tan(math.radians(self.targetAngleZ)))*(-1) + self.config['CAMERA_OFFSET_Y'] + self.config['NOTE_DIAMETER']/2
+        return self.targetErrorY
+
+    def streamVideo(self):
```
Comment (abdelazer) [inline notedetector.py:97]: Delete?

File: robot.py [hunk 21]
```diff
@@ -318,21 +334,53 @@ def teleopInit(self):
         return
 
     def robotPeriodic(self):
+        #print(self.mechanism.indexBeamBroken())
+        if self.notedetector.hasTarget():
+            print('target at ({}, {}) at {} degrees'.format(self.notedetector.getTargetErrorX(), self.notedetector.getTargetErrorY(), self.notedetector.getTargetErrorAngle()))
+        else:
+            #print('no target')
```
Comment (abdelazer) [inline robot.py:341]: Delete extra commented out code.

File: robot.py [hunk 22]
```diff
@@ -318,21 +334,53 @@ def teleopInit(self):
...
+            #print(self.notedetector.getCounter())
+            pass
         self.mechanism.periodic()
-        if not self.mechanism.indexBeamBroken():
-            LEDs.rainbowLED("purple")
+        if self.mechanism.indexBeamBroken():
+            LEDs.rainbowLED("purple-flash")
+            #print('purple-flash')
+        elif self.notedetector.hasTarget():
+            if self.notedetector.getTargetErrorX() < 6 and self.notedetector.getTargetErrorX() > -6:
```
Comment (abdelazer) [inline robot.py:350]: Pull these magic numbers (6, -6, 1.5) into variables or CONSTANTS

File: robot.py [hunk 23]
```diff
@@ -318,21 +334,53 @@ def teleopInit(self):
...
+
         field = wpilib.Field2d()
         field.setRobotPose(self.swervometer.getPathPlannerPose())
         self.elastic.putNumber('Match Time', wpilib.Timer.getMatchTime())
-        #self.elastic.putField(field)
         return True
-    
+
     def teleopPeriodic(self):
+        print(self.mechanism.shootingMotorRPMs)
```
Comment (abdelazer) [inline robot.py:379]: Do we want this print() enabled?

File: robot.py [hunk 24]
```diff
@@ -353,15 +401,15 @@ def teleopMechanism(self):
         if not self.mechanism.indexBeamBroken():
             self.mechanism.indexNote()
             self.mechanism.stopShooting()
-            LEDs.rainbowLED("purple")
+            #LEDs.rainbowLED("purple")
```
Comment (abdelazer) [inline robot.py:404]: Are we sure this should be commented out?

File: robot.py [hunk 25]
```diff
@@ -535,17 +583,24 @@ def teleopDrivetrain(self):
             self.drivetrain.setWheelLock(True)
         else:
             self.drivetrain.setWheelLock(False)
-        
+        if(driver.getAButton()):
```
Comment (abdelazer) [inline robot.py:586]: Nit: Space after `if`

File: robot.py [hunk 26]
```diff
@@ -535,17 +583,24 @@ def teleopDrivetrain(self):
...
+            self.drivetrain.alignWithApril(0, 75, 0)
```
Comment (abdelazer) [inline robot.py:587]: Pull out magic numbers?

File: robotconfig.py [hunk 27]
```diff
@@ -137,135 +137,27 @@
...
-                               ['RAISE_ARM_START', 0],
-                               ['PATH', '1-2'],
-                               ['SHOOT_NOTE'],
-                               ['LOWER_ARM_START', -37],
-                               ['PATH', '2-3'],
-                               ['RAISE_ARM_START', -25.9],
-                               ['PATH', '3-B'],
-                               ['SHOOT_NOTE']],
-
+    'B_FOUR_NOTE_GRAB': [['START_INTAKE'], ['RAISE_ARM_START', -25.9], ['WAIT', 1], ['SHOOT_NOTE'], ['LOWER_ARM_START', -37], ['PATH_TO_NOTE', 'B-1', 0.5], ['RAISE_ARM_START', -10], ['PATH', '1-SHOT'], ['SHOOT_NOTE'], ['LOWER_ARM_START', -37], ['PATH_TO_NOTE', 'SHOT-2', 0.3], ['RAISE_ARM_START', 0], ['WAIT', 0.5], ['SHOOT_NOTE'], ['LOWER_ARM_START', -37], ['PATH_TO_NOTE', '2-3', 0.3], ['RAISE_ARM_START', -5], ['PATH', '3-2'], ['SHOOT_NOTE'], ['LOWER_ARM_START', -37]],
```
Comment (abdelazer) [inline robotconfig.py:140]: I'll reformat this again after this PR is merged, so don't worry about it.

File: swervedrive.py [hunk 28]
```diff
@@ -167,6 +169,17 @@ def __init__(
 
         # TODO: 
         # - tune PID values
+        self.noteDrive_config = None
+        self.noteDrive_x_pid_controller = PIDController(0.025, 0 ,0)
```
Comment (abdelazer) [inline swervedrive.py:173]: Lots of magic numbers, but that's kinda the point of these, so potentially OK

File: swervedrive.py [hunk 29]
```diff
@@ -167,6 +169,17 @@ def __init__(
 
         # TODO: 
         # - tune PID values
+        self.noteDrive_config = None
```
Comment (abdelazer) [inline swervedrive.py:172]: Do you ever use this?

File: swervedrive.py [hunk 30]
```diff
@@ -701,6 +714,41 @@ def goToPose(self, x, y, bearing):
...
+        # for telemetry
+        currentX, currentY, currentBearing = self.swervometer.getCOF()
+        self.pose_target_x = currentX + (x * math.cos(currentBearing))
+        self.pose_target_y = currentY + (self.pose_target_x * math.tan(currentBearing + math.atan(y/x)))
+        self.pose_target_bearing = currentBearing + bearing
+
+        x_error = self.target_x_pid_controller.calculate(currentX, self.pose_target_x)
+        y_error = -self.target_y_pid_controller.calculate(currentY, self.pose_target_y)
+        
+        #self.log("hello: x: ", self.target_x_pid_controller.getSetpoint(), " y: ", self.target_y_pid_controller.getSetpoint())
```
Comment (abdelazer) [inline swervedrive.py:732]: Delete

File: swervedrive.py [hunk 31]
```diff
@@ -1092,6 +1140,29 @@ def alignWithApril(self, offsetX, offsetY, offsetAngle):
             self.set_rcw(0)
             self.execute('center')
     
+    def alignWithNote(self, offsetX, offsetY, offsetAngle):
+        x, y, r = self.swervometer.getCOF()
+
+        if(self.notedetector.hasTarget()):
+            if not offsetX is None:
```
Comment (abdelazer) [inline swervedrive.py:1147]: Nit: Python programmers would typically write this as `if offsetX is not None:`, but it's totally equivalent.

**Review by abdelazer — APPROVED** [review]
@ldavidc I'd suggest addressing my questions in a future PR against main (if you and @Cryplo think they're valid) rather than waiting on merging this PR.

File: autonomous.py (hunk 1 above)
Comment (abdelazer) [reply autonomous.py:76]: Dylan says this is intentional, as it's _relative_

File: autonomous.py (hunk 2 above)
Comment (abdelazer) [reply autonomous.py:79]: Same, this is the _relative_ mode

File: autonomous.py (hunk 3 above)
Comment (ldavidc) [reply autonomous.py:128]: Yes, that's a good idea. It's the maximum distance before jumping to alignWithNote.

File: autonomous.py (hunk 2 above)
Comment (ldavidc) [reply autonomous.py:79]: Also, relativeMove is not functional right now. Maybe we comment it out?

File: coralscript.py (hunk 4 above)
Comment (ldavidc) [reply coralscript.py:10]: It doesn't appear to be true anymore.

File: autonomous.py (hunk 5 above)
Comment (ldavidc) [reply autonomous.py:229]: We're planning on using presets instead of auto-aiming during auton

File: autonomous.py (hunk 6 above)
Comment (ldavidc) [reply autonomous.py:247]: POINT_TO_NOTE was for if we wanted to rotate, which we're not doing anymore, so we can comment it out or just delete it.

File: notedetector.py (hunk 13 above)
Comment (ldavidc) [reply notedetector.py:11]: Yes, streaming video isn't working.

File: robot.py (hunk 24 above)
Comment (ldavidc) [reply robot.py:404]: Yes. Lights are called in robotPeriodic now, not teleop only

**Pull request merged**